sklearn/ensemble/_hist_gradient_boosting/binning.py

```diff
@@ -16,76 +16,96 @@
-    binning_thresholds: list of ndarray
+    binning_thresholds: ndarray
         For each feature, stores the increasing numeric values that can
         be used to separate the bins. Thus ``len(binning_thresholds) ==
         n_features``.
     """
-    rng = check_random_state(random_state)
-    if subsample is not None and data.shape[0] > subsample:
-        subset = rng.choice(data.shape[0], subsample, replace=False)
-        data = data.take(subset, axis=0)
-
-    binning_thresholds = []
-    for f_idx in range(data.shape[1]):
-        col_data = data[:, f_idx]
-        # ignore missing values when computing bin thresholds
-        missing_mask = np.isnan(col_data)
-        if missing_mask.any():
-            col_data = col_data[~missing_mask]
-        col_data = np.ascontiguousarray(col_data, dtype=X_DTYPE)
-        distinct_values = np.unique(col_data)
-        if len(distinct_values) <= max_bins:
-            midpoints = distinct_values[:-1] + distinct_values[1:]
-            midpoints *= .5
-        else:
-            # We sort again the data in this case. We could compute
-            # approximate midpoint percentiles using the output of
-            # np.unique(col_data, return_counts) instead but this is more
-            # work and the performance benefit will be limited because we
-            # work on a fixed-size subsample of the full data.
-            percentiles = np.linspace(0, 100, num=max_bins + 1)
-            percentiles = percentiles[1:-1]
-            midpoints = np.percentile(col_data, percentiles,
-                                      interpolation='midpoint').astype(X_DTYPE)
-            assert midpoints.shape[0] == max_bins - 1
+    # ignore missing values when computing bin thresholds
+    missing_mask = np.isnan(col_data)
+    if missing_mask.any():
+        col_data = col_data[~missing_mask]
+    col_data = np.ascontiguousarray(col_data, dtype=X_DTYPE)
+    distinct_values = np.unique(col_data)
+    if len(distinct_values) <= max_bins:
+        midpoints = distinct_values[:-1] + distinct_values[1:]
+        midpoints *= .5
+    else:
+        # We sort again the data in this case. We could compute
+        # approximate midpoint percentiles using the output of
+        # np.unique(col_data, return_counts) instead but this is more
+        # work and the performance benefit will be limited because we
+        # work on a fixed-size subsample of the full data.
+        percentiles = np.linspace(0, 100, num=max_bins + 1)
+        percentiles = percentiles[1:-1]
+        midpoints = np.percentile(col_data, percentiles,
+                                  interpolation='midpoint').astype(X_DTYPE)
+        assert midpoints.shape[0] == max_bins - 1
+
+    # We avoid having +inf thresholds: +inf thresholds are only allowed in
+    # a "split on nan" situation.
+    np.clip(midpoints, a_min=None, a_max=ALMOST_INF, out=midpoints)
+    return midpoints
+
+
+def _find_bin_categories(col_data, max_bins, feature_idx):
```

Comment:
I'm not sure we need a new function here.
Why not just rely on `_find_binning_threshold`? One benefit would be that `bin_thresholds` would actually be consistent between numerical and categorical features.

Reply:
This is the same as `_find_binning_threshold`, until the error checking with the cardinality and error checking the encoded values. I do see a good way to combine them without making it more complicated.

Reply:
The categorical version would skip the whole midpoint finding or percentile piece of `_find_binning_threshold`.

Reply:
I think it would help to have a consistent `bin_thresholds` attribute. It would also make binning missing values faster. I'll give it a try.

Reply:
Also it would allow users to have encoded categories <= 255, and not just <= max_bins (though that's not a big deal anyway)

Reply:
OK so I tried that and indeed this doesn't seem to be really worth it, so I'll resolve. Also if we merge https://github.com/thomasjpfan/scikit-learn/pull/31 we don't even need this function anymore